### ISSUE TITLE
Crash apparently involving Nayru's and learning Sun's Song?

### ISSUE THREAD

**Issue by r0bd0g** [OoTRandomizer/OoT-Randomizer#1089]
https://clips.twitch.tv/ZanyAgreeableTrayPanicBasket

Occurred in this seed:
https://www.dropbox.com/s/tmlx7qqtab10j60/OoT_RSL_09062020.zpf?dl=0

Might this be vanilla?
Could the shortened cutscenes have allowed two things to overlap that normally wouldn't maybe?
This seed, composer grave was in goron city grotto - you could use Nayru's to get across the lava.

**Comment by ETR-BTF** [comment]
Crashes on gz too. Seems like a vanilla bug.

**Comment by mzxrules** [comment]
Seems likely it's the classic conflict with overlapping magic effects, not the easiest to fix.

**Comment by djevangelia** [comment]
Yes, vanilla bug. Nayru's Love is allocated in same space as magic/song effects. The problem is the two display lists in Nayru's Love (sDiamondMaterialDL, sDiamondModelDL) when song effect is spawned. For some reason, if Nayru's Love is killed and song effect spawned on the same frame, there is a risk of crash _even if neither are actually drawn that frame!!_ I'm not knowledgeable enough to understand or investigate why.

However, there seems to be an easy and seamless fix. The function is here: https://github.com/zeldaret/oot/blob/main/src/code/z_message.c#L3364-L3378
Nayru will be killed if message mode is `MSGMODE_OCARINA_CORRECT_PLAYBACK` or `MSGMODE_SONG_PLAYED` (see z_magic_dark.c). It seems that `MSGMODE_OCARINA_CORRECT_PLAYBACK` is used when not learning songs. The timer `msgCtx->stateTimer` is set to 10 when that mode is set, before moving to next mode. That gives time for Nayru to be killed. When learning songs, `MSGMODE_SONG_PLAYBACK_SUCCESS` seems to be used. This sets timer to 20, but played state has not been reached yet. When timer reaches zero, `MSGMODE_SONG_PLAYED` is set. Now, `Message_ContinueTextbox()` sets timer to 3, but a few lines down it's set to 1.
So one frame later, in `MSGMODE_SONG_PLAYED` when timer reaches 0, the Ocarina effects actor is spawned. Nayru is killed on the same frame as effect spawn and has a high risk of crashing.

So, the solution should be simply to increase the timer to 2 or higher (or remove the `msgCtx->stateTimer = 1;`), so that `MSGMODE_SONG_PLAYED` lasts longer than a frame.
I tried timer = 2 on NTSC 1.0 and MQ debug without crashing. Here's the NTSC run: https://www.youtube.com/watch?v=7HaHrXDBGGc (SoT cutscene removes Nayru so skipped that)
